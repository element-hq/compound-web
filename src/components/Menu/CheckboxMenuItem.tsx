/*
 * Copyright 2025 New Vector Ltd
 *
 * SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
 * Please see LICENSE files in the repository root for full details.
 */

import React, {
  type ComponentProps,
  forwardRef,
  useCallback,
  useId,
} from "react";
import { MenuItem } from "./MenuItem";
import { CheckboxInput } from "../Form";

type Props = Pick<
  ComponentProps<typeof MenuItem>,
  "className" | "label" | "onSelect" | "disabled"
> & {
  /**
   * Whether the checkbox is checked.
   */
  checked: boolean;
};

/**
 * A menu item with a checkbox control.
 * Must be used within a compound Menu or other `menu` or `menubar` aria role subtree.
 */
export const CheckboxMenuItem = forwardRef<HTMLInputElement, Props>(
  function CheckboxMenuItem(
    { className, label, onSelect, checked, disabled },
    ref,
  ) {
    const toggleId = useId();
    // The checkbox is controlled and we intend to ignore its events. We do need
    // to at least set onChange though to make React happy.
    const onChange = useCallback(() => {}, []);

    // <label> elements are not allowed to have a role like menuitemcheckbox, so
    // we must instead use a plain <div> for the menu item and use aria-checked
    // etc. to communicate its state.
    return (
      <MenuItem
        as="div"
        role="menuitemcheckbox"
        aria-checked={checked}
        className={className}
        label={label}
        onSelect={onSelect}
        disabled={disabled}
        Icon={
          <CheckboxInput
            id={toggleId}
            ref={ref}
            // This is purely cosmetic; really the whole MenuItem is the toggle.
            aria-hidden
            checked={checked}
            disabled={disabled}
            onChange={onChange}
          />
        }
      />
    );
  },
);
