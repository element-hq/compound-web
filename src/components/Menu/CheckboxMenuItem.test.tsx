/*
 * Copyright 2025 New Vector Ltd
 *
 * SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
 * Please see LICENSE files in the repository root for full details.
 */

import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import React from "react";
import userEvent from "@testing-library/user-event";
import { CheckboxMenuItem } from "./CheckboxMenuItem.tsx";

describe("CheckboxMenuItem", () => {
  it("renders", () => {
    const { asFragment } = render(
      <CheckboxMenuItem
        label="Always show"
        checked={false}
        onSelect={() => {}}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("toggles", async () => {
    const user = userEvent.setup();
    const toggle = vi.fn();
    render(
      <CheckboxMenuItem
        label="Always show"
        checked={false}
        onSelect={toggle}
      />,
    );

    // Try checking using keyboard controls
    await user.tab();
    await user.keyboard("[Space]");
    expect(toggle).toBeCalledTimes(1);
    toggle.mockClear();

    // Try checking by clicking
    await user.click(screen.getByRole("menuitemcheckbox"));
    expect(toggle).toBeCalledTimes(1);
    toggle.mockClear();

    // Also try clicking the label specifically (as this has regressed in the
    // past)
    await user.click(screen.getByText("Always show"));
    expect(toggle).toBeCalledTimes(1);
  });
});
