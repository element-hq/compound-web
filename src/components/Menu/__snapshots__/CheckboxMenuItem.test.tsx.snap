// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CheckboxMenuItem > renders 1`] = `
<DocumentFragment>
  <button
    aria-checked="false"
    class="_item_9618f6 _interactive_9618f6"
    data-kind="primary"
    role="menuitemcheckbox"
  >
    <div
      class="_container_937330 _icon_9618f6"
    >
      <input
        aria-hidden="true"
        class="_input_937330"
        id="«r0»"
        tabindex="-1"
        type="checkbox"
      />
      <div
        class="_ui_937330"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.55 17.575q-.2 0-.375-.062a.9.9 0 0 1-.325-.213L4.55 13q-.274-.274-.262-.713.012-.437.287-.712a.95.95 0 0 1 .7-.275q.425 0 .7.275L9.55 15.15l8.475-8.475q.274-.275.713-.275.437 0 .712.275.275.274.275.713 0 .437-.275.712l-9.2 9.2q-.15.15-.325.212a1.1 1.1 0 0 1-.375.063"
          />
        </svg>
      </div>
    </div>
    <span
      class="_typography_61e060 _font-body-md-medium_61e060 _label_9618f6"
    >
      Always show
    </span>
  </button>
</DocumentFragment>
`;
